File: LeetCode/JavaScript/0977-Squares-of-a-Sorted-Array.js
// 977. Squares of a Sorted Array
// https://leetcode.com/problems/squares-of-a-sorted-array/

/**
 * @param {number[]} nums
 * @return {number[]}
 */

var sortedSquares = function (nums) {
  const n = nums.length;
  const result = new Array(n);

  let left = 0;
  let right = n - 1;
  let pos = n - 1;

  while (left <= right) {
    const leftSquare = nums[left] * nums[left];
    const rightSquare = nums[right] * nums[right];

    if (leftSquare > rightSquare) {
      result[pos] = leftSquare;
      left++;
    } else {
      result[pos] = rightSquare;
      right--;
    }
    pos--;
  }

  return result;
};

File: LeetCode/Python/0977_Squares_of_a_Sorted_Array.py
# 977. Squares of a Sorted Array
# https://leetcode.com/problems/squares-of-a-sorted-array/


class Solution:
    def sortedSquares(self, nums: list[int]) -> list[int]:
        n = len(nums)
        result = [0] * n

        left = 0
        right = n - 1
        pos = n - 1

        while left <= right:
            left_square = nums[left] * nums[left]
            right_square = nums[right] * nums[right]

            if left_square > right_square:
                result[pos] = left_square
                left += 1
            else:
                result[pos] = right_square
                right -= 1

            pos -= 1

        return result
